python/pyarrow/array.pxi

```diff
@@ -988,13 +988,51 @@ cdef class Array(_PandasConvertible):
     def nbytes(self):
         """
         Total number of bytes consumed by the elements of the array.
+
+        In other words, the sum of bytes from all buffer 
+        ranges referenced.
+
+        Unlike `get_total_buffer_size` this method will account for array
+        offsets.
+
+        If buffers are shared between arrays then the shared
+        portion will be counted multiple times.
+
+        The dictionary of dictionary arrays will always be counted in their 
+        entirety even if the array only references a portion of the dictionary.
         """
-        size = 0
-        for buf in self.buffers():
-            if buf is not None:
-                size += buf.size
+        cdef:
+            shared_ptr[CArray] shd_ptr_c_array
+            CArray *c_array
+            CResult[int64_t] c_res_buffer
+
+        shd_ptr_c_array = pyarrow_unwrap_array(self)
+        c_array = shd_ptr_c_array.get()
```

Comment:
To get the C++ pointer, you can actually use `self.sp_array` or `self.ap` attributes, instead of calling `pyarrow_unwrap_array` on `self` (that method is only used for eg variables passed to a method)

Reply:
So I _think_ you can simplify this to `c_res_buffer = ReferencedBufferSize(deref(self.ap))`

Reply:
Yes, I also thought about that, but still used that API. :) 

I will modify the corresponding calls in rest of the functions.